app/schema/schema.go

```diff
@@ -139,15 +144,118 @@ func Action_GoCodegen(args *cli.Context) error {
 	outputDir := args.Path("output")
 	pkgName := args.String("package")
 
-	if generator != "go-gengo" {
+	switch generator {
+	case "go-gengo":
+		if err := os.MkdirAll(outputDir, os.ModePerm); err != nil {
+			return err
+		}
+		a := gengo.AdjunctCfg{}
+		gengo.Generate(outputDir, pkgName, ts, &a)
+	case "go-bindnode":
+		if err := generateGoBindnode(schemaFilePath, outputDir, pkgName, &ts); err != nil {
+			return err
+		}
+	default:
 		return fmt.Errorf("unsupported generator: %s", generator)
 	}
 
+	return nil
+}
+
+func generateGoBindnode(schemaFilePath, outputDir, pkgName string, ts *schema.TypeSystem) error {
 	if err := os.MkdirAll(outputDir, os.ModePerm); err != nil {
 		return err
 	}
-	a := gengo.AdjunctCfg{}
-	gengo.Generate(outputDir, pkgName, ts, &a)
 
-	return nil
+	// generate the basic Go types in types.go
+	f, err := os.Create(filepath.Join(outputDir, "types.go"))
+	if err != nil {
+		return err
+	}
+
+	if _, err := fmt.Fprintf(f, "package %s\n\n", pkgName); err != nil {
+		return err
+	}
+
+	delete(ts.GetTypes(), "Link")
+	ts.GetTypes()["Bogus"] = schema.SpawnBool("Bogus")
+	copy(ts.Names()[5:], ts.Names()[6:])
+	ts.Names()[len(ts.Names())-1] = "Bogus"
+
+	if err := bindnode.ProduceGoTypes(f, ts); err != nil {
+		return err
+	}
+
+	if err := f.Close(); err != nil {
+		return err
+	}
+
+	// generate schema prototypes in schema.go
+	relPath, err := filepath.Rel(outputDir, schemaFilePath)
+	if err != nil {
+		// TODO: better err
+		return err
+	}
+
+	type tmplfillIn struct {
+		PkgName         string
+		SchemaEmbedPath string
+		TypeNames       []string
+	}
+
+	tmpl, err := template.New("schematypegen").Parse(`
+package {{.PkgName}}
+
+import (
+	_ "embed"
+	
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/node/bindnode"
+	"github.com/ipld/go-ipld-prime/schema"
+)
+
+//go:embed {{.SchemaEmbedPath}}
+var embeddedSchema []byte
+
+var Types schemaSlab
+
+type schemaSlab struct {
+{{range .TypeNames}}
+{{.}}	schema.TypedPrototype{{end}}
+}
+
+func init() {
+	ts, err := ipld.LoadSchemaBytes(embeddedSchema)
+	if err != nil {
+		panic(err)
+	}
+{{range .TypeNames}}
+
+	Types.{{.}} = bindnode.Prototype(
+		(*{{.}})(nil),
+		ts.TypeByName("{{.}}"),
+	){{end}}
+}
+`)
+	if err != nil {
+		return err
+	}
+
+	buf := new(bytes.Buffer)
+	fill := &tmplfillIn{
+		PkgName:         pkgName,
+		SchemaEmbedPath: relPath,
+		TypeNames:       ts.Names()[5 : len(ts.Names())-1], // Skip basic types and the bogus type
```

Comment:
bindnode codegen indeed skips over "prelude" types, because they are handled with already-existing types. I personally think gengo should do the same - by using basicnode implementations and such.

As for the `5`, I think what you presumably want is the length of the prelude. We shouldn't really be hard-coding that. But I think we should more easily fix it by making gengo not generate code for prelude types.

Reply:
If gengo does need to generate code for prelude types, then I think the answer is that we should let gengo generate for those. I don't really know gengo closely enough to tell for sure. @warpfork will know.

Reply:
Yeah, I would hope that no action is needed here, and the generator should own the choice of not generating any content that it doesn't need to.

(Sidenote, I softcore think the bindnode generator should emit `type String string` aliases, derpy though that may look in golang, just for completeness and consistency of rules, because it's essentially free (afaik), aside from looking funny.  But that's tangential to whether anything gets filtered up here, and also a weakly held opinion.)